src/Features/RemotePaymentMethods/PaymentGatewaysController.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/**
+ * Logic for extending WC_REST_Payment_Gateways_Controller.
+ */
+
+namespace Automattic\WooCommerce\Admin\Features\RemotePaymentMethods;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * PaymentGateway class
+ */
+class PaymentGatewaysController {
+
+	/**
+	 * Initialize payment gateway changes.
+	 */
+	public static function init() {
+		add_filter( 'woocommerce_rest_prepare_payment_gateway', array( __CLASS__, 'extend_response' ), 10, 3 );
+	}
+
+	/**
+	 * Add necessary fields to REST API response.
+	 *
+	 * @param  WP_REST_Response   $response   Response data.
+	 * @param  WC_Payment_Gateway $gateway    Payment gateway object.
+	 * @param  WP_REST_Request    $request    Request object.
+	 * @return WP_REST_Response
+	 */
+	public static function extend_response( $response, $gateway, $request ) {
+		$data = $response->get_data();
+
+		$data['needs_setup'] = $gateway->needs_setup();
+
+		if ( method_exists( $gateway, 'get_oauth_connection_url' ) ) {
+			$data['oauth_connection_url'] = $gateway->get_oauth_connection_url();
+		}
+
+		$data['setup_fields'] = self::get_setup_form_fields( $gateway );
+
+		$data['settings_url'] = admin_url( 'admin.php?page=wc-settings&tab=checkout&section=' . strtolower( $gateway->id ) );
+
+		$response->set_data( $data );
+
+		return $response;
+	}
+
+	/**
+	 * Get setup fields from keys.
+	 *
+	 * @param  WC_Payment_Gateway $gateway    Payment gateway object.
+	 * @return array
+	 */
+	public static function get_setup_form_fields( $gateway ) {
+		if ( ! method_exists( $gateway, 'get_setup_form_field_keys' ) ) {
+			return apply_filters( 'woocommerce_settings_api_setup_form_fields_' . $gateway->id, array(), $gateway );
+		}
+
+		$setup_form_field_keys = $gateway->get_setup_form_field_keys();
+		$form_fields           = $gateway->get_form_fields();
+		$setup_fields          = array();
+
+		foreach ( $setup_form_field_keys as $key ) {
+			if ( ! isset( $form_fields[ $key ] ) ) {
+				continue;
+			}
+
+			$setup_fields[ $key ] = $form_fields[ $key ];
+		}
+
```

Comment:
What do you think about just returning the array of keys? In the response we have duplicate data and the properties under `setup_fields` isn't exhaustive. In the client we may need to look in two places to get all the data. Instead we could use the keys and loop through `settings` to get the data.

**setup_flelds**

<img width="616" alt="Screen Shot 2021-05-04 at 9 22 10 PM" src="https://user-images.githubusercontent.com/1922453/116984150-0e461f00-ad1f-11eb-8631-cb459d0a5334.png">

**settings**

<img width="570" alt="Screen Shot 2021-05-04 at 9 22 23 PM" src="https://user-images.githubusercontent.com/1922453/116984160-11410f80-ad1f-11eb-85bb-9bb2a6235b6d.png">

Reply:
>Note that I returned the full settings field data as opposed to just returning the keys. The reason is to allow further flexibility if a gateway wanted to use the filter woocommerce_settings_api_setup_form_fields_ to add/modify fields.

Just saw this after writing the above. This makes sense, just wondering why a gateway would filter these fields for setup so it differs from the settings screen. I suppose different messaging would be useful for a user's initial setup v. updating an existing value. In the end I think this is solving a problem that may or may not exist. Unless I'm missing a use case, I prefer simple but won't stand in the way of it either. What do you think?

Reply:
>Unless I'm missing a use case, I prefer simple but won't stand in the way of it either. What do you think?

My thought was that custom text snippet fields could be used here.

But on further thought, I think this can be hacked around using the keys and if custom content is really needed, they can always SlotFill.

Updated this to just return the keys and not the fields.